moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,1254 @@
+      {
+        planning_scene.processCollisionObjectMsg(kv.second);
+      }
+
+      collision_result.clear();
+      planning_scene.checkCollision(collision_request, collision_result);
+
+      // Scale robot velocity according to collision proximity and user-defined
+      // thresholds.
+      // I scaled exponentially (cubic power) so velocity drops off quickly
+      // after the threshold.
+      // Proximity decreasing --> decelerate
+      double velocity_scale = 1;
+
+      // Ramp velocity down linearly when collision proximity is between
+      // lower_collision_proximity_threshold and
+      // hard_stop_collision_proximity_threshold
+      if ((collision_result.distance > parameters.hard_stop_collision_proximity_threshold) &&
+          (collision_result.distance < parameters.lower_collision_proximity_threshold))
+      {
+        // scale = k*(proximity-hard_stop_threshold)^3
+        velocity_scale =
+            64000. * pow(collision_result.distance - parameters.hard_stop_collision_proximity_threshold, 3);
+      }
+      else if (collision_result.distance < parameters.hard_stop_collision_proximity_threshold)
+        velocity_scale = 0;
+
+      velocity_scale = velocity_scale_filter.filter(velocity_scale);
+      // Put a ceiling and a floor on velocity_scale
+      if (velocity_scale > 1)
+        velocity_scale = 1;
+      else if (velocity_scale < 0.05)
+        velocity_scale = 0.05;
+
+      // Very slow if actually in collision
+      if (collision_result.collision)
+        velocity_scale = 0.02;
+
+      pthread_mutex_lock(&shared_variables.collision_velocity_scale_mutex);
+      shared_variables.collision_velocity_scale = velocity_scale;
+      pthread_mutex_unlock(&shared_variables.collision_velocity_scale_mutex);
+
+      collision_rate.sleep();
+    }
+  }
+}
+
+// Constructor for the class that handles jogging calculations
+JogCalcs::JogCalcs(const JogArmParameters& parameters, jog_arm_shared& shared_variables,
+                   const std::unique_ptr<robot_model_loader::RobotModelLoader>& model_loader_ptr)
+  : move_group_(parameters.move_group_name), tf_listener_(tf_buffer_)
+{
+  parameters_ = parameters;
+
+  // Publish collision status
+  warning_pub_ = nh_.advertise<std_msgs::Bool>(parameters_.warning_topic, 1);
+
+  // MoveIt Setup
+  // Wait for model_loader_ptr to be non-null.
+  while (ros::ok() && !model_loader_ptr)
+  {
+    ROS_WARN_THROTTLE_NAMED(5, NODE_NAME, "Waiting for a non-null robot_model_loader pointer");
+    ros::Duration(0.1).sleep();
+  }
+  const robot_model::RobotModelPtr& kinematic_model = model_loader_ptr->getModel();
+  kinematic_state_ = std::make_shared<robot_state::RobotState>(kinematic_model);
+  kinematic_state_->setToDefaultValues();
+
+  joint_model_group_ = kinematic_model->getJointModelGroup(parameters_.move_group_name);
+
+  std::vector<double> dummy_joint_values;
+  kinematic_state_->copyJointGroupPositions(joint_model_group_, dummy_joint_values);
```

Comment:
are you assuming you are initializing the joint group to zero value? what is the point of these two lines?

Reply:
Yep, I guess that doesn't make sense. I removed the lines and there seems to be no effect.